src/Closure.cpp

```diff
@@ -106,5 +109,79 @@ void Closure::visit(const Atomic *op) {
     op->body.accept(this);
 }
 
+Expr Closure::pack_into_struct() const {
+    std::vector<Expr> elements;
+    std::vector<Type> types;
+
+    std::ostringstream name;
+
+    for (const auto &b : buffers) {
+        Expr ptr_var = Variable::make(type_of<void *>(), b.first);
+        elements.emplace_back(ptr_var);
+    }
+    for (const auto &v : vars) {
+        Expr var = Variable::make(v.second, v.first);
+        elements.emplace_back(var);
+    }
+
+    // Sort by decreasing size, to guarantee the struct is densely packed in
+    // memory. We don't actually rely on this, it's just nice to have.
+    std::stable_sort(elements.begin(), elements.end(),
+                     [&](const Expr &a, const Expr &b) {
+                         return a.type().bytes() > b.type().bytes();
+                     });
+
+    for (Expr e : elements) {
+        types.push_back(e.type());
+        name << e.type() << "_";
+    }
+
+    Expr result = Call::make(Handle(),
+                             Call::make_struct, elements, Call::Intrinsic);
+    return result;
+}
+
+Stmt Closure::unpack_from_struct(const Expr &e, const Stmt &s) const {
+    // Use the struct-packing code just to make sure the order of elements is
```

Comment:
Since this code relies on `Closure` having a deterministic order for vars and buffers, we should add a comment to that effect Closure.h in case someone ever wants to change the usage of `std::map` to (eg) `std::unordered_map`.

Reply:
It's the same closure object, so the order would have to change between two iterations over the same map. I don't think even unordered_map is that non-deterministic.